src/fido2_child/fido2_child_credentials.c

```diff
@@ -0,0 +1,295 @@
+    }
+
+    ret = sss_generate_csprng_buffer(userid, sizeof(userid));
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_generate_csprng_buffer failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Setting user: %s.\n", data->shortname);
+
+    ret = fido_cred_set_user(cred, userid, sizeof(userid), data->shortname,
+                             NULL, NULL);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_user failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = fido_cred_set_rk(cred, FIDO_OPT_FALSE);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_set_rk failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+ssize_t
+read_pin(char **pin)
+{
+    char *line_ptr = NULL;
+    struct termios old, new;
+    size_t line_len = 0;
+    ssize_t bytes_read;
+    ssize_t ret;
+
+    ret = tcgetattr(STDIN_FILENO, &old);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unable get the parameters associated with stdin.\n");
+        goto done;
+    }
+    new = old;
+    new.c_lflag &= ~ECHO;
+    ret = tcsetattr(STDIN_FILENO, TCSAFLUSH, &new);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE, "Unable to turn echoing off.\n");
+        goto done;
+    }
+
+    printf("Enter PIN: ");
+    bytes_read = getline(&line_ptr, &line_len, stdin);
+    if (bytes_read == -1) {
+        DEBUG(SSSDBG_OP_FAILURE, "getline failed [%d]: %s.\n",
+              errno, sss_strerror(errno));
+    } else {
+        // Remove the end of line '\n' character
+        line_ptr[--bytes_read] = '\0';
+    }
+    printf("\n");
+
+    ret = tcsetattr(STDIN_FILENO, TCSAFLUSH, &old);
+    if (ret != 0) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unable to restore parameters associated with stdin.\n");
+        goto done;
+    }
+
+    ret = bytes_read;
+    *pin = line_ptr;
+
+done:
+    return ret;
+}
+
+errno_t
+generate_credentials(struct fido2_data *data, fido_dev_t *dev,
+                     fido_cred_t *cred)
+{
+    char *pin = NULL;
+    bool has_pin;
+    bool has_uv;
+    ssize_t pin_len;
+    errno_t ret;
+
+    has_uv = fido_dev_has_uv(dev);
+    has_pin = fido_dev_has_pin(dev);
+    if (data->user_verification == FIDO_OPT_TRUE && has_uv == false
+        && has_pin == false) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Policy enabled user-verification but there isn't any "
+              "verification method set.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    if (has_uv == true) {
+        fido_cred_set_uv(cred, FIDO_OPT_TRUE);
+    }
+
+    if (has_pin == true) {
+        pin_len = read_pin(&pin);
+        if (pin_len == -1) {
+            ret = ERR_INPUT_PARSE;
+            goto done;
+        }
+    }
+
+    ret = fido_dev_make_cred(dev, cred, pin);
```

Comment:
When a PIN is used for user verification, you also need to call `fido_cred_set_uv(cred, FIDO_OPT_TRUE)` after `fido_dev_make_cred()` if you want `fido_cred_verify{,_self}()` to check that the UV flag was set in the authenticator response.

Reply:
I see, thank you